pkg/snapshot/clonetree.go

```diff
@@ -0,0 +1,363 @@
+					return err
+				}
+				err = createPartitionDir(buildPartitionDir, srcPartitionDir)
+				if err != nil {
+					return err
+				}
+			}
+		} else if fi.Mode().IsRegular() {
+			// Regular files in the block device directory are both regular and
+			// pseudofiles containing information such as the size (in sectors)
+			// and whether the device is read-only
+			buf, err := ioutil.ReadFile(fp)
+			if err != nil {
+				return err
+			}
+			targetPath := filepath.Join(buildDeviceDir, fname)
+			trace("creating %s\n", targetPath)
+			f, err := os.Create(targetPath)
+			if err != nil {
+				return err
+			}
+			if _, err = f.Write(buf); err != nil {
+				return err
+			}
+			f.Close()
+		}
+	}
+	// There is a special file $DEVICE_DIR/queue/rotational that, for some hard
+	// drives, contains a 1 or 0 indicating whether the device is a spinning
+	// disk or not
+	srcQueueDir := filepath.Join(
+		srcDeviceDir,
+		"queue",
+	)
+	buildQueueDir := filepath.Join(
+		buildDeviceDir,
+		"queue",
+	)
+	err = os.MkdirAll(buildQueueDir, os.ModePerm)
+	if err != nil {
+		return err
+	}
+	fp := filepath.Join(srcQueueDir, "rotational")
+	buf, err := ioutil.ReadFile(fp)
+	if err != nil {
+		return err
+	}
+	targetPath := filepath.Join(buildQueueDir, "rotational")
+	trace("creating %s\n", targetPath)
+	f, err := os.Create(targetPath)
+	if err != nil {
+		return err
+	}
+	if _, err = f.Write(buf); err != nil {
+		return err
+	}
+	f.Close()
+
+	return nil
+}
+
+func createPartitionDir(buildPartitionDir string, srcPartitionDir string) error {
+	// Populate the supplied directory (in our build filesystem) with all the
+	// appropriate information pseudofile contents for the partition.
+	partFiles, err := ioutil.ReadDir(srcPartitionDir)
+	if err != nil {
+		return err
+	}
+	for _, f := range partFiles {
+		fname := f.Name()
+		fp := filepath.Join(srcPartitionDir, fname)
+		fi, err := os.Lstat(fp)
+		if err != nil {
+			return err
+		}
+		if fi.Mode()&os.ModeSymlink != 0 {
+			// Ignore any symlinks in the partition directory since they simply
+			// point to information we aren't interested in like "subsystem"
+			continue
+		} else if fi.IsDir() {
+			// The subdirectories in the partition directory are not
+			// interesting for us. They have information about power events and
+			// traces
+			continue
+		} else if fi.Mode().IsRegular() {
+			// Regular files in the block device directory are both regular and
+			// pseudofiles containing information such as the size (in sectors)
+			// and whether the device is read-only
+			buf, err := ioutil.ReadFile(fp)
+			if err != nil {
+				return err
+			}
+			targetPath := filepath.Join(buildPartitionDir, fname)
+			trace("creating %s\n", targetPath)
+			f, err := os.Create(targetPath)
+			if err != nil {
+				return err
+			}
+			if _, err = f.Write(buf); err != nil {
+				return err
+			}
+			f.Close()
+		}
+	}
+	return nil
+}
+
+type CopyFileOptions struct {
+	IsSymlink func(path string, info os.FileInfo) bool
```

Comment:
docstring for this attribute please. Also, perhaps we might rename this to "IsSymlinkFunction" or "IsSymlinkFn"?

Reply:
you might include in the docstring an indication of usage, too.

Reply:
good idea, done.